substrate/frame/delegated-staking/src/lib.rs

```diff
@@ -0,0 +1,763 @@
+		// transferred to actual delegator.
+		let proxy_delegator = Self::sub_account(AccountType::ProxyDelegator, who.clone());
+
+		// Keep proxy delegator alive until all funds are migrated.
+		frame_system::Pallet::<T>::inc_providers(&proxy_delegator);
+
+		// Get current stake
+		let stake = T::CoreStaking::stake(who)?;
+
+		// release funds from core staking.
+		T::CoreStaking::migrate_to_virtual_staker(who);
+
+		// transferring just released staked amount. This should never fail but if it does, it
+		// indicates bad state and we abort.
+		T::Currency::transfer(who, &proxy_delegator, stake.total, Preservation::Expendable)
+			.map_err(|_| Error::<T>::BadState)?;
+
+		T::CoreStaking::update_payee(who, reward_account)?;
+		Self::do_delegate(&proxy_delegator, who, stake.total)
+	}
+
+	fn do_bond(agent_acc: &T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
+		let agent = Agent::<T>::get(agent_acc)?;
+
+		let available_to_bond = agent.available_to_bond();
+		defensive_assert!(amount == available_to_bond, "not expected value to bond");
+
+		if agent.is_bonded() {
+			T::CoreStaking::bond_extra(&agent.key, amount)
+		} else {
+			T::CoreStaking::virtual_bond(&agent.key, amount, agent.reward_account())
+		}
+	}
+
+	fn do_delegate(
+		delegator: &T::AccountId,
+		agent: &T::AccountId,
+		amount: BalanceOf<T>,
+	) -> DispatchResult {
+		let mut ledger = AgentLedger::<T>::get(agent).ok_or(Error::<T>::NotAgent)?;
+		// try to hold the funds.
+		T::Currency::hold(&HoldReason::StakingDelegation.into(), delegator, amount)?;
+
+		let new_delegation_amount =
+			if let Some(existing_delegation) = Delegation::<T>::get(delegator) {
+				ensure!(&existing_delegation.agent == agent, Error::<T>::InvalidDelegation);
+				existing_delegation
+					.amount
+					.checked_add(&amount)
+					.ok_or(ArithmeticError::Overflow)?
+			} else {
+				amount
+			};
+
+		Delegation::<T>::new(agent, new_delegation_amount).update_or_kill(delegator);
+		ledger.total_delegated =
+			ledger.total_delegated.checked_add(&amount).ok_or(ArithmeticError::Overflow)?;
+		ledger.update(agent);
+
+		Self::deposit_event(Event::<T>::Delegated {
+			agent: agent.clone(),
+			delegator: delegator.clone(),
+			amount,
+		});
+
+		Ok(())
+	}
+
+	fn do_release(
+		who: &T::AccountId,
+		delegator: &T::AccountId,
+		amount: BalanceOf<T>,
+		num_slashing_spans: u32,
+	) -> DispatchResult {
+		let mut agent = Agent::<T>::get(who)?;
+		let mut delegation = Delegation::<T>::get(delegator).ok_or(Error::<T>::NotDelegator)?;
+
+		// make sure delegation to be released is sound.
+		ensure!(&delegation.agent == who, Error::<T>::NotAgent);
+		ensure!(delegation.amount >= amount, Error::<T>::NotEnoughFunds);
+
+		// if we do not already have enough funds to be claimed, try withdraw some more.
+		if agent.ledger.unclaimed_withdrawals < amount {
+			// withdraw account.
+			let _ = T::CoreStaking::withdraw_unbonded(who.clone(), num_slashing_spans)
```

Comment:
Maybe we save the return value of this as `stash_killed`...

Reply:
There are two steps for a withdrawal:
- Agent Withdraw: Funds are withdrawn from staking ledger but not claimed yet by delegator.
- Delegator withdraw: This is when funds are actually released in delegator account.

The above represents `Agent Withdraw` which can kill the stash in staking pallet. The delegation staking pallet needs to keep the agent until funds are claimed by a delegator. Which is why we ignore the return value for `withdraw_unbonded`.

These two steps generally happens in sync but not always. If the staking ledger is filled up with `MaxUnlockingChunks`, it will do an implicit withdraw where these calls become async.